api/filrewardsd/service/service.go

```diff
@@ -0,0 +1,571 @@
+		cacheOrg[rec.DevKey][rec.Type] = struct{}{}
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("iterating cursor while building cache: %v", err)
+	}
+
+	s := &Service{
+		rewardsCol:        rewardsCol,
+		claimsCol:         claimsCol,
+		rewardsCacheOrg:   cacheOrg,
+		rewardsCacheDev:   cacheDev,
+		baseAttoFILReward: config.BaseAttoFILReward,
+		semaphores:        nutil.NewSemaphorePool(1),
+	}
+
+	if config.AnalyticsAddr != "" {
+		s.ac, err = analytics.New(config.AnalyticsAddr)
+		if err != nil {
+			return nil, fmt.Errorf("creating analytics client: %s", err)
+		}
+	}
+
+	s.server = grpc.NewServer()
+	go func() {
+		pb.RegisterFilRewardsServiceServer(s.server, s)
+		if err := s.server.Serve(config.Listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Errorf("serve error: %v", err)
+		}
+	}()
+
+	return s, nil
+}
+
+func (s *Service) ProcessAnalyticsEvent(ctx context.Context, req *pb.ProcessAnalyticsEventRequest) (*pb.ProcessAnalyticsEventResponse, error) {
+	if req.OrgKey == "" {
+		return nil, status.Error(codes.InvalidArgument, "must provide org key")
+	}
+	if req.AnalyticsEvent == analyticspb.Event_EVENT_UNSPECIFIED {
+		return nil, status.Error(codes.InvalidArgument, "must provide analytics event")
+	}
+
+	lck := s.semaphores.Get(orgKeyLock(req.OrgKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	t := pb.RewardType_REWARD_TYPE_UNSPECIFIED
+	switch req.AnalyticsEvent {
+	case analyticspb.Event_EVENT_KEY_ACCOUNT_CREATED:
+		t = pb.RewardType_REWARD_TYPE_FIRST_KEY_ACCOUNT_CREATED
+	case analyticspb.Event_EVENT_KEY_USER_CREATED:
+		t = pb.RewardType_REWARD_TYPE_FIRST_KEY_USER_CREATED
+	case analyticspb.Event_EVENT_ORG_CREATED:
+		t = pb.RewardType_REWARD_TYPE_FIRST_ORG_CREATED
+	case analyticspb.Event_EVENT_BILLING_SETUP:
+		t = pb.RewardType_REWARD_TYPE_INITIAL_BILLING_SETUP
+	case analyticspb.Event_EVENT_BUCKET_CREATED:
+		t = pb.RewardType_REWARD_TYPE_FIRST_BUCKET_CREATED
+	case analyticspb.Event_EVENT_BUCKET_ARCHIVE_CREATED:
+		t = pb.RewardType_REWARD_TYPE_FIRST_BUCKET_ARCHIVE_CREATED
+	case analyticspb.Event_EVENT_MAILBOX_CREATED:
+		t = pb.RewardType_REWARD_TYPE_FIRST_MAILBOX_CREATED
+	case analyticspb.Event_EVENT_THREAD_DB_CREATED:
+		t = pb.RewardType_REWARD_TYPE_FIRST_THREAD_DB_CREATED
+	}
+	if t == pb.RewardType_REWARD_TYPE_UNSPECIFIED {
+		// It is normal to get an analytics event we aren't interested in, so just return an empty result and no error.
+		return &pb.ProcessAnalyticsEventResponse{}, nil
+	}
+
+	ensureKeyRewardCache(s.rewardsCacheOrg, req.OrgKey)
+	ensureKeyRewardCache(s.rewardsCacheDev, req.DevKey)
+
+	if _, exists := s.rewardsCacheOrg[req.OrgKey][t]; exists {
+		// This reward is already granted to the Org so bail.
+		return &pb.ProcessAnalyticsEventResponse{}, nil
+	}
+
+	if _, exists := s.rewardsCacheDev[req.DevKey][t]; exists {
+		// This reward is already granted to the Dev so bail.
+		return &pb.ProcessAnalyticsEventResponse{}, nil
+	}
```

Comment:
@andrewxhill here is problem with the js client not sending dev key when using org api key... we'd like to be able to check that the dev has not unlocked the reward for _any_ org, but if we don't have a dev key, we can't do this check here.

Reply:
I've changed my thinking about this. Basically, if a dev wants to use an org API key in the js client, that org key is free to unlock any rewards that haven't been previously unlocked for that org. It's like a bonus for the user doing something advanced like using the js client.

Reply:
wouldn't that would allow me to create unlimited orgs for unlimited rewards

Reply:
Each org can only be rewarded each reward type once no matter who/where the reward type was unlocked from. So it's not unlimited for one org. But yes, you could create unlimited number of orgs and unlock all rewards for each of them using the org API key. This is the problem I was surfacing in my original comment above.

Reply:
The only way to prevent that is to somehow not grant fil rewards unless we have the dev key _and_ org key.

Reply:
We could try some of the follow mitigations now or in the future:
- Require that an org be of a certain age before any rewards are unlocked (1 week or similar).
- Require that an org receiving rewards be owned by a dev that has not received awards, directly and/or via a different org they own.

Reply:
Nice ideas. The org age thing is nice. Thoughts on that @andrewxhill?

The second bullet is basically the way it already works... A reward can only be unlocked if the dev hasn't unlocked the same reward for any/all orgs they belong to. And the reward can only be unlocked if it hasn't already been unlocked for the org.